### DISCUSSION TITLE
Question about the difference between the song cache and downloaded songs

### DISCUSSION THREAD

**Discussion by matchboxbananasynergy** [z-huang/music#940] [Q&A] [answered]
Hey there! I've got a quick question about how InnerTune functions with regards to offline playback.

From what I can see, there is currently a song cache (which I assume works the same way ViMusic's equivalent did), but you also have the ability to download songs manually, which don't seem to be downloaded as actual files that are accessible by other apps, but rather just downloaded to the app's internal storage.

Is my assumption correct that the difference between these two features and the reason for having both is that the song cache caches songs for offline playback as you listen to them, whereas with downloads you can "pre-cache" songs without first having to listen to them? I suppose that with downloads, you can make sure that songs/albums are *kept* offline, whereas with the cache, if it's filled up, they start disappearing and being replaced with others.

Also, regarding the cache, does it work like ViMusic's in the sense that when it's filled up, it starts getting rid of the songs that you haven't listened to the longest first?

Thanks in advance for your reply; just trying to make sure I fully understand how the app works so that I can introduce it to other folks appropriately. :)

**Comment by z-huang** [comment, marked as answer]
You're assumption is correct. The cache is same as ViMusic's cache, following the least recently used (LRU)  strategy. Downloads can be viewed as cache, too, but with unlimited quota. The data flow is: internet -> cache -> download cache -> player. The player searches for song in the download cache and the cache first. Only if the song doesn't exist in both cache will the player fetch data from the internet.
You may notice that some downloads are super fast. That's because if the song is cached, the app just makes a copy from the cache to the download cache.

**Reply by owenconnorz** [reply]
Downloads are awesome @z-huang hope there will be a fix for the no stream available I'm not 100% sure is it possible that the issues maybe the following.

When you changed the cookies for the login.

Maybe a api issue.

I'm not 100% but now login has been added is piped able to be removed now as we can use login to verify our age.
